java/ql/lib/semmle/code/java/PrettyPrintAst.qll

```diff
@@ -770,6 +784,36 @@ private class PpSwitchCase extends PpAst, SwitchCase {
   }
 }
 
+private class PpPatternCase extends PpAst, PatternCase {
+  override string getPart(int i) {
+    i = 0 and result = "case "
+    or
+    i = 2 and this.getPattern() instanceof LocalVariableDeclExpr and result = " "
+    or
+    i = 3 and result = this.getPattern().asBindingPattern().getName()
+    or
+    i = 2 + this.getPatternOffset() and result = ":" and not this.isRule()
+    or
+    i = 2 + this.getPatternOffset() and result = " -> " and this.isRule()
+    or
+    i = 4 + this.getPatternOffset() and result = ";" and exists(this.getRuleExpression())
+  }
+
+  private int getPatternOffset() {
+    if this.getPattern() instanceof LocalVariableDeclExpr then result = 2 else result = 0
+  }
+
+  override PpAst getChild(int i) {
+    i = 1 and result = this.getPattern().asBindingPattern().getTypeAccess()
+    or
+    i = 1 and result = this.getPattern().asRecordPattern()
+    or
+    i = 4 and result = this.getRuleExpression()
```

Comment:
This index is too low - it clashes with the `" -> "` in `getPart` when this is a simple pattern.

Reply:
I guess both this line and the disjunct below should simply be `i=5`.

Reply:
And then we might as well make the `":"` and the `" -> "` be at index 4 unconditionally and delete the `getPatternOffset` predicate.